assembly/assembly-root-war/src/main/webapp/_app/oauthLoader.js

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+import { KeycloakLoader } from './keycloackLoader.js';
+
+/**
+ * Displays error message
+ * @param {string} message an error message to show
+ */
+function error(message) {
+    const container = document.getElementById("error-console");
+
+    const element = document.createElement("pre");
+    element.innerHTML = message;
+    element.style.color = 'red';
+    container.appendChild(element);
+    return element;
+}
+
+/**
+ * Returns an array of query parameter values if it is present
+ * @param {string} name of the query parameter
+ */
+function getQueryParams(name) {
+    const queryString = window.location.search;
+    return new URLSearchParams(queryString).getAll(name);
+}
+
+/**
+ * Fetches userId
+ */
+function asyncGetUserId() {
+    return new Promise((resolve, reject) => {
+        const request = new XMLHttpRequest();
+        request.open("GET", '/api/user');
+        setAuthorizationHeader(request).then((xhr) => {
+            xhr.send();
+            xhr.onreadystatechange = () => {
+                if (xhr.readyState !== 4) {
+                    return;
+                }
+                if (xhr.status !== 200) {
+                    const errorMessage = 'Failed to get the workspace: "' + this.getRequestErrorMessage(xhr) + '"';
+                    reject(new Error(errorMessage));
+                    return;
+                }
+                resolve(JSON.parse(xhr.responseText).id);
+            };
+        });
+    });
+}
+
+/**
+ * Sets authorization header for a request
+ * @param {XMLHttpRequest} xhr
+ */
+function setAuthorizationHeader(xhr) {
+    return new Promise((resolve, reject) => {
+        if (window._keycloak && window._keycloak.token) {
+            window._keycloak.updateToken(5).success(() => {
+                xhr.setRequestHeader('Authorization', 'Bearer ' + window._keycloak.token);
+                resolve(xhr);
+            }).error(() => {
+                window.sessionStorage.setItem('oidcIdeRedirectUrl', location.href);
+                window._keycloak.login();
+                reject(new Error('Failed to refresh token'));
+            });
+        }
+
+        resolve(xhr);
+    });
+}
+
+function sendToken(token) {
+    if (window.opener) {
```

Comment:
it would be reasonable to check if `window.opener` points to the workspace that belongs to the current user.

Reply:
I can't read the opener's url. It is blocked by browser restrictions: 
```
Exception: DOMException: Blocked a frame with origin "http://che-che.192.168.99.253.nip.io" from accessing a cross-origin frame. at sendToken (http://che-che.192.168.99.253.nip.io/_app/oauthLoader.js:86:26)<error: SecurityError: Blocked a frame with origin "http://che-che.192.168.99.253.nip.io" from accessing a cross-origin frame.<error> at http://che-che.192.168.99.253.nip.io/_app/oauthLoader.js:132:13]
```

Reply:
I added a check that compares machine token received from parent window and the keycloak token: https://github.com/eclipse/che/blob/2e31d1208863509be770d0ce8cee91dcedf18025/assembly/assembly-root-war/src/main/webapp/_app/oauthLoader.js#L123-L143